admin/Dockerfile.qa

```diff
@@ -29,12 +29,37 @@ apt-get install -y sudo \
 
 # Install Python 3.8 as the default "python3" executable on older linux distribution
 apt install -y python3-setuptools
-if [ "$DOCKER_OS" = "ubuntu:18.04" -o "$DOCKER_OS" = "ubuntu:20.04" ]; then
+if [ "$DOCKER_OS" = "ubuntu:18.04" ]; then
+  # cspell:disable
+  apt install -y wget \
+                 build-essential \
+                 checkinstall \
+                 libreadline-gplv2-dev \
+                 libncursesw5-dev \
+                 libssl-dev \
+                 libsqlite3-dev \
+                 tk-dev \
+                 libgdbm-dev \
+                 libc6-dev \
+                 libbz2-dev \
+                 libffi-dev \
+                 zlib1g-dev
+  # cspell:enable
+  cd /tmp
+  PythonVersion=3.9.20
+  wget https://www.python.org/ftp/python/$PythonVersion/Python-$PythonVersion.tgz
+  tar xzf Python-$PythonVersion.tgz
+  cd Python-$PythonVersion
+  ./configure --enable-optimizations
+  # cspell:ignore altinstall
+  make altinstall
+  update-alternatives --install /usr/local/bin/python3 python3 /usr/local/bin/python3.9 10
```

Comment:
This part is absolutely awful but I'm glad you found a way.

I have a minor concern that installing the `build-essential` package implies a GCC version that might differ from the one we request with `$CC`/`$CXX`, but that shouldn't be able to interfere.

An alternative to building Python ourselves could be download an ancient version from some binary archive. For example, the `uv` package manager (alternative to `pip`) downloads pre-built Pythons from the https://github.com/indygreg/python-build-standalone project. But doing that requires extra platform detection (e.g. aarch64 vs x86_64) and a step to unpack the archives.

Reply:
I removed the build essentials and it still works., using our gcc version. It would be hard to build the filename for a prebuilt version and building it on our own doesn't take much time. The build for the whole container takes about 160s.
I played with the layering to first build the common parts and then install GCC and Python but the time benefit is worse against the effort for maintaining such a Dockerfile.

Reply:
We can improve this in a separate PR and maybe build the Python in a separate stage `FROM ... AS PythonBuild` and then install it in the final container.